torch/_inductor/codegen/cpp.py

```diff
@@ -4019,6 +4019,100 @@ def can_fuse_vertical(self, node1, node2):
+        """
+        # No reduction and no mudular
+        if any(
+            len(node.group[1][1]) != 0
+            or any(
+                expr.has(ModularIndexing) for expr in node._body.indexing_exprs.values()
+            )
+            for node in nodes
+        ):
+            return nodes
+
+        split_var = None
+        split_number = None
+        divide_index_name = None
+
+        def has_one_div(node):
+            num_div = 0
+            expr_matched = False
+            for name, expr in node._body.indexing_exprs.items():
+                num_div += expr.count(FloorDiv)
+                if expr.count(FloorDiv) == 1:
+                    nonlocal split_var
+                    nonlocal split_number
+                    nonlocal divide_index_name
+                    div_expr = expr.find(FloorDiv).pop()
+                    split_var = div_expr.args[0]
+                    split_number = div_expr.args[1]
+                    divide_index_name = name
+                    expr_matched = (
+                        isinstance(split_number, sympy.core.numbers.Integer)
+                        and isinstance(split_var, sympy.core.symbol.Symbol)
+                        and divide_index_name is not None
+                    )
+
+            return num_div == 1 and expr_matched
+
+        # Only one node contains a division, and the split dimension is contiguous in all other indexing_exprs.
+        nodes_have_div = [node for node in nodes if has_one_div(node)]
+        if len(nodes_have_div) != 1 or any(
+            stride_at_vec_range(
+                expr, split_var, cpu_vec_isa.pick_vec_isa().nelements(torch.float32)
+            )
+            != 1
+            for name, expr in nodes_have_div[0]._body.indexing_exprs.items()
+            if name != divide_index_name
+        ):
+            return nodes
+
+        for node in nodes:
+            iter_ranges = [*node._body.var_ranges.values()]
+            iter_vars = [*node._body.var_ranges.keys()]
+            split_idx = iter_vars.index(split_var)
+            assert split_idx < len(iter_vars)
+            new_iter_ranges: Tuple[sympy.Expr, ...] = ()
+            for idx in range(len(iter_ranges)):
+                if idx != split_idx:
+                    new_iter_ranges += (iter_ranges[idx],)
+                else:
+                    val_to_split = iter_ranges[idx]
+                    new_iter_ranges += (val_to_split // split_number, split_number)
+
+            new_index = sympy.core.symbol.Symbol(
+                "z" + str(len(iter_ranges)), integer=True, nonnegative=True
+            )
+            new_iter_vars = [
+                var * split_number + new_index if var == split_var else var
+                for var in iter_vars
+            ]
+
+            # Here decide the final loop order
+            node.recompute_size_and_body(
+                extra_size_and_var_constraints=(
```

Comment:
Here I extend recompute_size_and_body. Instead of passing in the index expression and then running the logic in simplify_and_reorder to get vars and ranges, I pass in vars and ranges via the arg `extra_size_and_var_constraints` and use them directly to re-trace the LoopBody.

Reply:
Oh, I meant we can reuse the existing `extra_indexing_contraints` without introducing another constraints here.

Reply:
Do you mean reusing`extra_indexing_contraints ` to pass vars and ranges?